### ISSUE TITLE
[ENHANCEMENT] Unclear role system vs user authentication with "on-behalf" applications

### ISSUE THREAD

**Issue by stdweird** [Arize-ai/phoenix#4980]
**Describe the bug**

(I am not sure this is a feature request or a bug. It's not cery clear from the docs if this a use case for the system key. I am afraid not, which makes this a feature request ;) )

a. When deploying e.g. a web chatbot with SSO authentication, are we supposed to use a system key as authentication for the traces? The docs seem to suggest not to, as system keys have a very "admin"-like description, and i don't fully trust each web application reponsible with all admin powers of phoenix ;)

b. Similar, but slightly different. Assume in the webapp above, we pass "username"  as metadata for each trace, corresponding with the authenticated user in the webapp at the time of request. If we have the same SSO provider configured with phoenix, is there a way that the traces of this application can be seen by 
- a group of users (authenticated in phoenix) "responsible" for the web application. (I think what i am asking is something like a service user (and key) that is mapped to group of real users (the responsibles for the application). This service credential can then be used to deploy the app and record the traces.)
- by an authenticated user whose username matches the username metadata (this user would ofcourse only have access to traces matching its username).

**Comment by mikeldking** [comment]
Hey @stdweird, hope I can provide some clarity

> a. When deploying e.g. a web chatbot with SSO authentication, are we supposed to use a system key as authentication for the traces? The docs seem to suggest not to, as system keys have a very "admin"-like description, and i don't fully trust each web application reponsible with all admin powers of phoenix ;)

For deployments of applications we'd recommend using the system keys stored in secrets. These keys are managed by admins and not visible to pure members of Phoenix. While they do have full access to Phoenix, this is by design for us as we'd like you to be able to programmatically send anything you need from your application runtime. At some time in the future I think we might add the idea of "scope" for keys but this is currently planned.


I think for the second inquiry, totally understand the desire to manage access to projects and traces for users. Right now we don't have this either as we want teams to be able to fully collaborate on everything. We definitely will have to re-visit this design choice similar to above where users only have access to sub-parts of the UI but right now we assume small teams collaborate fully on a single instance of phoenix.

Thanks for the feedback. We definitely keep your use-case in mind as we enhance key access and RBAC.

**Comment by stdweird** [comment]
hey @mikeldking . thanks for the reply. so at this moment:
- the developers/responsibles of this webchat application form a team, and teams can be isolated from other teams.
- the endusers of the webchat are not in the picture of the authentication (or we would have to make them all member of the team; but that is a bit far fetched). 
- the user keys are then also to be used for single user teams; multi user teams should use a system key and the phoenix admins have to trust them with this power. 

does that sum up the current state? 
i looked a bit further in the docs how to define a team linked to a project, but i don't seem to find any links to assigning a user to a team and/or project.  how does that work (or what did i overlook?)

**Comment by mikeldking** [comment]
@stdweird You are right in that Phoenix doesn't have the notion of teams just users. 

Here's an overview of the auth feature that might help https://youtu.be/e9I2voaGuQE?si=pss7Ly7b0WkW52p3

In general I'd advise the use of system keys for deployments and personal user keys for development on user machines.

If teams is super important to you definitely give app.arize.com a try. It has orgs and workspaces. Phoenix doesn't have a way to isolate access on a single instance at the moment. Sorry about that

**Comment by stdweird** [comment]
@mikeldking  thanks for the reply. i think we will settle on a single phoenix instance for single user usage; and investigate setting up a phoenix instance per project, so we have the granularity we need.
i will investigate the app.arize.com esp wrt hosting and possible legal issues. 

thanks for the effort!

**Comment by stdweird** [comment]
@mikeldking  sorry to reopen this, but there is another practical issue with having to run a phoenix instance per project: is there a way that based on SSO authentication, that we can control the actual access to phoenix. perhaps some python code we can inject as a plugin to filter out "valid" users to allow them access to the webinterface (and thus manage keys and view data etc).

we would deploy all the containers with same SSO setup, and then have some other mechanism (worst case an env var with some regex, or comma-separated list of user names) determine who gets access to what container. in an ideal scenario, we would have to register a separate app  for each project that has it's own container to control the authentication fully with SSO, but that is not a very light procedure on our end.

**Comment by mikeldking** [comment]
I’m currently out of the country so won’t be able to help with this unfortunately. I don’t think a email allowlist regex would be totally unreasonable. The dev team should be able to help via slack and other means. Thanks for the feedback!MikyoOn Nov 12, 2024, at 6:47 PM, Stijn De Weirdt ***@***.***> wrote:﻿
@mikeldking  sorry to reopen this, but there is another practical issue with having to run a phoenix instance per project: is there a way that based on SSO authentication, that we can control the actual access to phoenix. perhaps some python code we can inject as a plugin to filter out "valid" users to allow them access to the webinterface (and thus manage keys and view data etc).
we would deploy all the containers with same SSO setup, and then have some other mechanism (worst case an env var with some regex, or comma-separated list of user names) determine who gets access to what container. in an ideal scenario, we would have to register a separate app  for each project that has it's own container to control the authentication fully with SSO, but that is not a very light procedure on our end.

**Comment by JakobKallestad** [comment]
Hi, I am seconding this feature request. Is there any update on letting users "see their own traces, but not other peoples traces". I understand that the current approach is for all the traces to have shared access for the members within a project, but this can be challenging for our usecases.

If we have 50 different users (that we collaborate closely with) that use our chatbot and they will also give feedback in phoenix via annotations. Then we don't want to expose all the traces to all the users. This is because of privacy reasons, data access restrictions (if RAG with user-based access to data is being used), but also it just makes more sense to let users annotate and curate their own traces rather then being involved with other peoples traces, doesn't it?

Suggestion:
All traces can be tagged with user-ids based on SSO auth. Allow the user to log in using SSO to phoenix, but only let them see their own traces. 
Alternative solution --> If a user wan to provide feedback to a trace that has just occurred in the chatbot, then can we generate a direct link to let a user open this trace in phoenix to do annotations and curations, but not let the user see any other traces via that link?


Seeing as for our projects we are not able to have shared access with all the users of the tool, then our only real alternative if we want to use phoenix is to have a separate app or workflow that that will let users edit and curate traces before they are sendt to phoenix, but that essentially removes all the good features that phoenix offers to those users.

Thanks for reading the full comment. Hoping to get some feedback on this one.
